/*
 * This file is part of the NUbots Codebase.
 *
 * The NUbots Codebase is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * The NUbots Codebase is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright 2013 NUbots <nubots@nubots.net>
 */

syntax = "proto3";

package messages;

import "messages/Vector.proto";

message mat {
    repeated vec v = 1;
}

message fmat {
    repeated fvec v = 1;
}

message imat {
    repeated ivec v = 1;
}

message umat {
    repeated uvec v = 1;
}

message mat22 {
    vec2 x = 1;
    vec2 y = 2;
}

message fmat22 {
    fvec2 x = 1;
    fvec2 y = 2;
}

message imat22 {
    ivec2 x = 1;
    ivec2 y = 2;
}

message umat22 {
    uvec2 x = 1;
    uvec2 y = 2;
}

message mat33 {
    vec3 x = 1;
    vec3 y = 2;
    vec3 z = 3;
}

message fmat33 {
    fvec3 x = 1;
    fvec3 y = 2;
    fvec3 z = 3;
}

message imat33 {
    ivec3 x = 1;
    ivec3 y = 2;
    ivec3 z = 3;
}

message umat33 {
    uvec3 x = 1;
    uvec3 y = 2;
    uvec3 z = 3;
}

message mat44 {
    vec4 x = 1;
    vec4 y = 2;
    vec4 z = 3;
    vec4 t = 4;
}

message fmat44 {
    vec4 x = 1;
    vec4 y = 2;
    vec4 z = 3;
    vec4 t = 4;
}

message imat44 {
    vec4 x = 1;
    vec4 y = 2;
    vec4 z = 3;
    vec4 t = 4;
}

message umat44 {
    vec4 x = 1;
    vec4 y = 2;
    vec4 z = 3;
    vec4 t = 4;
}